### PR TITLE
COE-893: fix issue with GitHub action after merging unit test COE-810 branch

### PR DESCRIPTION
**Pull request by namnhatpham**

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -26,13 +29,20 @@ env:
 jobs:
   run-tests:
     runs-on: ubuntu-latest
-    environment: Nam
+    environment: test
```
Comment (mvoigt-sag) [inline .github/workflows/test.yml:29]: why do you need to define the environment again?

**Review by mvoigt-sag — CHANGES_REQUESTED** [review]
- please add @MarkJMReynolds-SAG as a reviewer to get his feedback
- in general I'm not sure if we really want to automatically create a release after merging something into main; I think we triggered the release build manually until now
- for the test execution: is it possible to get an overview of the executed/failed tests? similar to JUnit test overview?

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -6,33 +6,43 @@ on:
       environment:
         type: string
```
Comment (mvoigt-sag) [inline .github/workflows/test.yml:7]: - this can be improved by using `type: environment` => this will then show a dropdown box to select one of the existing environments

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -4,35 +4,45 @@ on:
...
-  C8Y_USER: ${{ secrets.C8Y_USER }}
+  C8Y_BASEURL: ${{ vars.C8Y_BASEURL }}
+  C8Y_TENANT: ${{ vars.C8Y_TENANT }}
+  C8Y_USER: ${{ vars.C8Y_USER }}
 
 jobs:
   run-tests:
     runs-on: ubuntu-latest
-    environment: Nam
+    environment: ${{ github.event.deployment.environment || github.event.inputs.environment }}
```
Comment (MarkJMReynolds-SAG) [inline .github/workflows/test.yml:32]: Again, if we're only running manually, don't need to choose between environments.

**Review by MarkJMReynolds-SAG — CHANGES_REQUESTED** [review]

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -4,35 +4,45 @@ on:
...
+        options:
+          oee-dev: oee-dev Environment
         required: true
       release_body:
         type: string
         description: GitHub Release Message
-        default: "GitHub Release"
+        default: "GitHub Manual Release"
 
   push:
```
[the remark this answers is not in the document]
Comment (mvoigt-sag) [reply .github/workflows/test.yml:18]: I agree to have a manual button push for the release process. But I'm not sure if we need to run tests for a release b/c we only build releases from the main branch and everything is tested (in a PR) before it comes to main...

[the remark this answers is not in the document]
Comment (MarkJMReynolds-SAG) [reply .github/workflows/test.yml:18]: I guess a lot of it comes down to how likely you think that scenario is.

[the remark this answers is not in the document]
Comment (mvoigt-sag) [reply .github/workflows/test.yml:18]: Correct. When thinking more about it: even a single PR could possibly break the tests in main after merging b/c the tests are only executed on the branch, not on the merged result ...

**Review by mvoigt-sag — APPROVED** [review]

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -2,25 +2,30 @@ name: Build oee-simulators
...
-
       - name: Update version number
         run: |
           cd repo/simulators
           SIMULATOR_VERSION=${GITHUB_RUN_NUMBER}
           sed -i "s/@project.version@/1.0.${SIMULATOR_VERSION}/g" cumulocity.json
-          echo "RELEASE_VERSION=${SIMULATOR_VERSION}" >> $GITHUB_ENV
+          echo "release_version=${SIMULATOR_VERSION}" >> $GITHUB_ENV
+        env:
+          release_version: ${{ env.release_version }}
```
Comment (mvoigt-sag) [inline .github/workflows/build.yml:28]: is that really needed? I dont see the value of `env.release_version` being used in the step ..

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -4,35 +4,37 @@ on:
...
+        type: environment
         description: Name of the environment to use
+        default: "oee-dev"
+        options:
+          oee-dev: oee-dev Environment
         required: true
       release_body:
         type: string
         description: GitHub Release Message
-        default: "GitHub Release"
```
Comment (mvoigt-sag) [inline .github/workflows/test.yml:13]: I think we dont need to specify the release body here since the release job should not be triggered from the test workflow...

**Review by mvoigt-sag — CHANGES_REQUESTED** [review]

File: .github/workflows/build.yml (hunk 5 above)
Comment (namnhatpham) [reply .github/workflows/build.yml:28]: Yes, it is passed to the release workflow

File: .github/workflows/build.yml [hunk 7]
```diff
@@ -1,26 +1,101 @@
...
+        type: bool
+
+  push:
+    branches: [ '*' ]
 
 jobs:
+  call-test-workflow:
+    uses: ./.github/workflows/test.yml
+    with:
+      environment: 'oee-dev'
```
Comment (mvoigt-sag) [inline .github/workflows/build.yml:38]: should not be a fixed version but the value from the input

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -1,26 +1,101 @@
...
+    secrets:
+      C8Y_PASSWORD: ${{ secrets.C8Y_PASSWORD }}
+
+  recall-test-workflow-on-other-environment:
```
Comment (mvoigt-sag) [inline .github/workflows/build.yml:42]: why is the test workflow called twice?

File: .github/workflows/build.yml [hunk 9]
```diff
@@ -1,26 +1,101 @@
...
+        description: GitHub Release Message
+        default: "Release version ${GITHUB_RUN_NUMBER}"
+
+      run_release:
+        description: 'Run release workflow?'
+        required: false
+        default: false
+        type: bool
+
+      run_build:
```
Comment (mvoigt-sag) [inline .github/workflows/build.yml:25]: is that input really required/used?

File: .github/workflows/test.yml [hunk 10]
```diff
@@ -1,50 +1,55 @@
...
-        description: GitHub Release Message
-        default: "GitHub Release"
 
-  push:
-    branches: [ '*' ]
-    environment:
-      name: Nam
+    secrets:
+      C8Y_PASSWORD:
+        required: true
```
Comment (mvoigt-sag) [inline .github/workflows/test.yml:12]: why is this required as input?

**Review by mvoigt-sag — CHANGES_REQUESTED** [review]

File: .github/workflows/test.yml [hunk 11]
```diff
@@ -1,50 +1,55 @@
...
+
+  workflow_dispatch:
```
[the remark this answers is not in the document]
Comment (namnhatpham) [reply .github/workflows/test.yml:14]: Yes, it can be if they are the same

File: .github/workflows/test.yml (hunk 10 above)
Comment (namnhatpham) [reply .github/workflows/test.yml:12]: Because the tests can not be run without password, I add required as True although it is not necessary

File: .github/workflows/build.yml (hunk 8 above)
Comment (namnhatpham) [reply .github/workflows/build.yml]: https://cumulocity.atlassian.net/browse/COE-910
I explained it in the new tasks that I created myself to test for a possibility

File: .github/workflows/test.yml (hunk 10 above)
Comment (mvoigt-sag) [reply .github/workflows/test.yml:12]: but what kind of password is that? where is it used? shouldnt we use the password from the environment?

File: .github/workflows/test.yml [hunk 12]
```diff
@@ -1,50 +1,55 @@
...
-  C8Y_USER: ${{ secrets.C8Y_USER }}
+  C8Y_BASEURL: ${{ vars.C8Y_BASEURL }}
+  C8Y_TENANT: ${{ vars.C8Y_TENANT }}
+  C8Y_USER: ${{ vars.C8Y_USER }}
 
 jobs:
   run-tests:
     runs-on: ubuntu-latest
-    environment: Nam
+    environment: ${{ github.event.inputs.environment || inputs.environment }}
```
Comment (mvoigt-sag) [inline .github/workflows/test.yml:33]: is github.event.inputs.environment still required here?

File: .github/workflows/test.yml [hunk 13]
```diff
@@ -1,50 +1,55 @@
...
     steps:
       - name: Checkout Repo
         uses: actions/checkout@v3
         with:
           path: 'repo'
 
+      - name: Check Environment
+        run:
+          echo "Running on environment ${{ github.event.inputs.environment || inputs.environment }}"
```
Comment (mvoigt-sag) [inline .github/workflows/test.yml:42]: is "github.event.inputs.environment" still required?

File: .github/workflows/build.yml (hunk 9 above)
Comment (mvoigt-sag) [reply .github/workflows/build.yml:25]: - I'd think that we call release.yml if we want to create a release; this will then call build.yml to create the binary
- I'd call release.yml directly if I want to create a release

File: .github/workflows/release.yml [hunk 14]
```diff
@@ -35,31 +39,28 @@ jobs:
           zip oee-simulators.zip image.tar cumulocity.json
```
Comment (mvoigt-sag) [inline .github/workflows/release.yml:39]: - not sure if that will work; where is the image.tar coming from?
- I'd expect a call to build.yml to create the binaries

**Review by mvoigt-sag — CHANGES_REQUESTED** [review]

File: .github/workflows/test.yml (hunk 10 above)
Comment (namnhatpham) [reply .github/workflows/test.yml:12]: Yes, the Password from environment is get in the build.yml and passed to this test.yml

File: .github/workflows/release.yml (hunk 14 above)
Comment (namnhatpham) [reply .github/workflows/release.yml:39]: This step in build.yml produce image.tar;
      - name: Build Docker image
        run: |
          cd repo/simulators
          docker build -t oee-simulators .
          docker save -o image.tar oee-simulators
          docker rmi oee-simulators

File: .github/workflows/test.yml [hunk 15]
```diff
@@ -1,50 +1,58 @@
...
+        required: true
+
+  workflow_dispatch:
+    inputs:
+      environment:
+        type: environment
         description: Name of the environment to use
+        default: "oee-dev"
+        options:
+          oee-dev: oee-dev Environment
```
Comment (mvoigt-sag) [inline .github/workflows/test.yml:21]: I think we dont need those options here

File: .github/workflows/test.yml [hunk 16]
```diff
@@ -1,50 +1,58 @@
...
-  C8Y_USER: ${{ secrets.C8Y_USER }}
+  C8Y_BASEURL: ${{ vars.C8Y_BASEURL }}
+  C8Y_TENANT: ${{ vars.C8Y_TENANT }}
+  C8Y_USER: ${{ vars.C8Y_USER }}
 
 jobs:
   run-tests:
     runs-on: ubuntu-latest
-    environment: Nam
+    environment: ${{ inputs.environment || 'oee-dev' }}
```
Comment (mvoigt-sag) [inline .github/workflows/test.yml:36]: ne need to specify 'oee-dev'

File: .github/workflows/test.yml [hunk 17]
```diff
@@ -1,50 +1,58 @@
...
     steps:
       - name: Checkout Repo
         uses: actions/checkout@v3
         with:
           path: 'repo'
 
+      - name: Check Environment
+        run:
+          echo "Running on environment ${{ inputs.environment || 'oee-dev' }}"
```
Comment (mvoigt-sag) [inline .github/workflows/test.yml:45]: remove 'oee-dev'

File: .github/workflows/test.yml (hunk 10 above)
Comment (namnhatpham) [reply .github/workflows/test.yml:12]: https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment#environment-secrets

Note: Workflows that run on self-hosted runners are not run in an isolated container, even if they use environments. Environment secrets should be treated with the same level of security as a repository and organization secrets. For more information, see "[Security hardening for GitHub Actions](https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#hardening-for-self-hosted-runners)."

**Review by mvoigt-sag — CHANGES_REQUESTED** [review]
